compiler+runtime/src/cpp/jank/read/lex.cpp

```diff
@@ -540,27 +737,95 @@ namespace jank::read
               return err(error{ token_start, pos, "unexpected end of real, expecting exponent" });
             }
 
+            if(expecting_more_digits)
+            {
+              ++pos;
+              if(found_r)
+              {
+                return err(error{ token_start,
+                                  pos,
+                                  "unexpected end of radix number, expecting more digits" });
+              }
+              return err(error{
+                token_start,
+                pos,
+                fmt::format("unexpected end of radix {} number, expecting more digits", radix) });
+            }
             /* Tokens beginning with - are ambiguous; it's only a negative number if it has numbers
              * to follow.
              * TODO: handle numbers starting with `+` */
             if(file[token_start] != '-' || (pos - token_start) >= 1)
             {
               require_space = true;
               ++pos;
-              if(contains_dot || is_scientific)
+              auto number_start{ token_start };
+              if(file[token_start] == '-' || file[token_start] == '+')
               {
-                return ok(token{ token_start,
-                                 pos - token_start,
-                                 token_kind::real,
-                                 std::strtod(file.data() + token_start, nullptr) });
+                number_start = token_start + 1;
+                found_beginning_negative = file[token_start] == '-';
+              }
+              if(found_r)
+              {
+                radix = std::stoi(file.data() + token_start, nullptr);
+                if(radix < 0)
+                {
+                  radix = -radix;
+                  found_beginning_negative = true;
+                }
+                if(radix < 2 || radix > 36)
+                {
+                  return err(
+                    error{ token_start,
+                           pos,
+                           fmt::format("invalid number: radix {} is out of range", radix) });
+                }
+                number_start = r_pos + 1;
+              }
+              else if(radix == 16)
+              {
+                number_start += 2;
               }
 
+              /* check for invalid digits */
+              native_vector<char> invalid_digits{};
+              for(auto i{ number_start }; i < pos; i++)
+              {
+                if(!is_valid_num_char(file[i], radix))
+                {
+                  invalid_digits.emplace_back(file[i]);
+                }
+              }
+              if(!invalid_digits.empty())
+              {
+                return err(
+                  error{ token_start,
+                         pos,
+                         fmt::format("invalid number: char {} are invalid for radix {}",
```

Comment:
Let's clarify the error message here and fix the grammar by saying `invalid number: chars '{}' are invalid for radix {}`.

Reply:
updated